config/keybindings.go

```diff
@@ -1,24 +1,216 @@
+			return
+		}
+		if strings.HasPrefix(binding, "Ctrl") {
+			m += tcell.ModCtrl
+		}
+	}
+
+	bindings[keyBinding{k, m, r}] = cmd
+}
+
+func KeyInit() {
+	configBindings := map[int]string{
+		CmdLink1:       "keybindings.bind_link1",
+		CmdLink2:       "keybindings.bind_link2",
+		CmdLink3:       "keybindings.bind_link3",
+		CmdLink4:       "keybindings.bind_link4",
+		CmdLink5:       "keybindings.bind_link5",
+		CmdLink6:       "keybindings.bind_link6",
+		CmdLink7:       "keybindings.bind_link7",
+		CmdLink8:       "keybindings.bind_link8",
+		CmdLink9:       "keybindings.bind_link9",
+		CmdLink0:       "keybindings.bind_link0",
+		CmdBottom:      "keybindings.bind_bottom",
+		CmdEdit:        "keybindings.bind_edit",
+		CmdHome:        "keybindings.bind_home",
+		CmdBookmarks:   "keybindings.bind_bookmarks",
+		CmdAddBookmark: "keybindings.bind_add_bookmark",
+		CmdSave:        "keybindings.bind_save",
+		CmdReload:      "keybindings.bind_reload",
+		CmdBack:        "keybindings.bind_back",
+		CmdForward:     "keybindings.bind_forward",
+		CmdPgup:        "keybindings.bind_pgup",
+		CmdPgdn:        "keybindings.bind_pgdn",
+		CmdNewTab:      "keybindings.bind_new_tab",
+		CmdCloseTab:    "keybindings.bind_close_tab",
+		CmdNextTab:     "keybindings.bind_next_tab",
+		CmdPrevTab:     "keybindings.bind_prev_tab",
+		CmdQuit:        "keybindings.bind_quit",
+		CmdHelp:        "keybindings.bind_help",
+		CmdSub:         "keybindings.bind_sub",
+		CmdAddSub:      "keybindings.bind_add_sub",
+	}
+	configTabNBindings := map[int]string{
+		CmdTab1: "keybindings.bind_tab1",
+		CmdTab2: "keybindings.bind_tab2",
+		CmdTab3: "keybindings.bind_tab3",
+		CmdTab4: "keybindings.bind_tab4",
+		CmdTab5: "keybindings.bind_tab5",
+		CmdTab6: "keybindings.bind_tab6",
+		CmdTab7: "keybindings.bind_tab7",
+		CmdTab8: "keybindings.bind_tab8",
+		CmdTab9: "keybindings.bind_tab9",
+		CmdTab0: "keybindings.bind_tab0",
+	}
+	tcellKeys = make(map[string]tcell.Key)
+	bindings = make(map[keyBinding]int)
+
+	for k, kname := range tcell.KeyNames {
+		tcellKeys[kname] = k
+	}
+
+	for c, allb := range configBindings {
+		for _, b := range viper.GetStringSlice(allb) {
+			parseBinding(c, b)
+		}
+	}
+
+	// Backwards compatibility with the old shift_numbers config line.
+	shiftNumbers := []rune(viper.GetString("keybindings.shift_numbers"))
+	if len(shiftNumbers) > 0 && len(shiftNumbers) <= 10 {
+		for i, r := range shiftNumbers {
+			bindings[keyBinding{tcell.KeyRune, 0, r}] = CmdTab1 + i
+		}
+	} else {
+		for c, allb := range configTabNBindings {
+			for _, b := range viper.GetStringSlice(allb) {
+				parseBinding(c, b)
 			}
-			return i + 1, nil
 		}
 	}
-	return -1, errors.New("provided key is invalid") //nolint:goerr113
+}
+
+func TranslateKeyEvent(e *tcell.EventKey) int {
+	var ok bool
+	var cmd int
+	k := e.Key()
+	if k == tcell.KeyRune {
+		cmd, ok = bindings[keyBinding{k, e.Modifiers(), e.Rune()}]
+	} else { // Sometimes tcell sets e.Rune() on non-KeyRune events.
```

Comment:
Where have you seen this? Just curious.

Reply:
Primarily on the Ctrl-Letter-Key events.  The event will have `e.Key() == KeyCtrlA`, `e.Modifers() == 2` (for Control) and `e.Rune() == 'A'`.  I originally added a correction for this when generating the bindings table, but it felt awkward, and would break if tcell changed behavior.